File: solutions/kotlin/grade-school/2/src/main/kotlin/GradeSchool.kt
private typealias Student = String
private typealias Grade = Int

class School {
    private val studentsByGrade = mutableMapOf<Grade, List<Student>>()

    fun add(student: Student, grade: Grade) {
        require(student !in roster())

        studentsByGrade[grade] = studentsIn(grade) + student
    }

    fun grade(grade: Grade): List<Student> = studentsIn(grade).sorted()

    fun roster(): List<Student> = studentsByGrade
        .toSortedMap()
        .values
        .flatMap { it.sorted() }

    private fun studentsIn(grade: Grade): List<Student> = studentsByGrade.getOrDefault(grade, emptyList())
}

File: solutions/fsharp/grade-school/2/GradeSchool.fs
module GradeSchool

type Grade = int

type Student = string

type School = Map<Grade, Student list>

let empty: School = Map.empty

let grade (number: Grade) (school: School): Student list =
    school |> Map.tryFind number |> Option.defaultValue []

let private containsStudentInAnyGrade student school =
    school |> Map.exists (fun _ students -> List.contains student students)

let add (student: Student) (target: Grade) (school: School): School =
    school
    |> containsStudentInAnyGrade student
    |> function
        | true -> school
        | _ ->
          let students = school |> grade target
          school |> Map.add target (student :: students |> List.sort)


let roster (school: School): Student list =
    school |> Map.toList |> List.collect snd
